src/logic/map_objects/tribes/soldier.cc

```diff
@@ -1774,9 +1774,9 @@ void Soldier::naval_invasion_update(Game& game, State& state) {
 		return;
 	}
 
-	// Camp is blocked, try to get as close as we can
+	// Camp is blocked, try to get as close as we can.
 	for (int i = 10; i > 0; --i) {
-		Coords coords = game.random_location(state.coords, kPortSpaceRadius);
+		Coords coords = game.random_location(state.coords, kPortSpaceRadius * kPortSpaceRadius);
```

Comment:
I was wrong, this is not enough. They never stop, and it even slows down the game very badly.

But I wouldn't want to increase the radius of the previous check (for conquering) unconditionally. Maybe separate the check for conquering from the check for stopping? It would also be nice if you could make them fill the immediate neighbours of the portspace first. (I'm sorry for coming up with more complications.)

Also, I apologise for nitpicking, but squaring the radius just isn't right logically.

Reply:
They may remain a bit scattered, just as I feared, but it's not that bad after all, and at least the code didn't have to get any more complicated. :)